[stable/swagger] add a chart for Swagger UI

jeff-knurek opened the pull request.

Commit:
[stable/swagger] add a chart for Swagger UI

jeff-knurek commented:
at the moment there is a soft dependency on `oauth-proxy` which is in `incubator` but in the process of merging to `stable`: https://github.com/kubernetes/charts/pull/3480

Commit:
SWAGGER: remove extra empty line at the end of the file

mattfarina commented on stable/swagger/templates/deployment.yaml:
```diff
@@ -0,0 +1,54 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
```
The `extensions/v1beta1` apiversion will be going away for deployments. It's late in the deprecation cycle. Could you use `apps/v1beta2` instead?

mattfarina commented:
/ok-to-test

mattfarina commented:
/hold

We need to hold on this until oauth-proxy is in stable. Stable charts cannot depend on incubator.

Commit:
SWAGGER: update deployment api to apps/v1beta2 (which depends on k8s 1.8+)

Commit:
SWAGGER: update oauth dependency to use stable oauth2-proxy chart

Commit:
SWAGGER: remove cert-manager as a direct dependency, it's still a good addition to this chart, but should be managed seperately

unguiculus commented:
Why does this chart have nginx-ingress and oauth-proxy as dependencies? An ingress controller usually has a broader scope. You don't install it on a per-chart basis. In order to support oauth-proxy, it might be better to allow sidecars to be configured. See #4283.

Commit:
SWAGGER: remove requirements file that contained nginx/oauth

Commit:
SWAGGER: change the documentation around using oauth2-proxy and nginx-ingress

jeff-knurek commented:
@unguiculus your point makes sense. `nginx-ingress` should be managed separately and not as a dependency of this chart. And by extension the `oauth2-proxy` doesn't make sense as a dependency either. 

Since the `oauth2-proxy` does a good job of managing any such containers, I don't see a need for adding a sidecar to this project either and creating more complexity when it could be obtained with another chart. But I'm open to more examples on how this could be a benefit for swagger.

Commit:
SWAGGER: fix typo

unguiculus commented:
Please compare with what you get with `helm create` using Helm 2.8+ and update the PR.

unguiculus commented:
/hold

Commit:
SWAGGER: update chart to use latest best practices from Helm 2.8.2 (also update the swagger api version)

jeff-knurek commented:
@unguiculus anything else needed?

unguiculus commented on stable/swagger/templates/deployment.yaml:
```diff
@@ -0,0 +1,66 @@
+            successThreshold: 1
+            timeoutSeconds: 2
+          ports:
+            - containerPort: {{ .Values.service.internalPort }}
```
You should probably hard-code the container port. Changing it is likely to break things. Use a named port and reference it by name from service and probes.

jeff-knurek replied:
I removed this setting from the README, but kept them as templated values.

I agree that `Changing it is likely to break things`, however the same can be said for pretty much every other value. If someone chooses to use a different swagger image that exposes a different port, than they can't use this helm chart? Just because the port is hard-coded? that doesn't make sense to me.

unguiculus replied:
Removing it from the readme doesn't make sense. If you want to have that feature, it should be documented. The question is whether we should really cater for the fact the someone might use a different port in their custom image. You could just say you don't support that. I don't see why you'd want to change the port anyways.

jeff-knurek replied:
I entirely disagree with your logic, but it's not worth blocking this chart from being available

unguiculus commented on stable/swagger/templates/service.yaml:
```diff
@@ -0,0 +1,19 @@
+    - port: {{ .Values.service.externalPort }}
+      targetPort: {{ .Values.service.internalPort }}
+      protocol: {{ .Values.service.protocol }}
+      name: {{ .Values.service.name }}
```
This is not the service name, it's the name of the port, which you'll probably want to hard-code.

unguiculus commented on stable/swagger/README.md:
```diff
@@ -0,0 +1,128 @@
+    kubernetes.io/ingress.class: nginx
+    kubernetes.io/tls-acme: 'true'
+  tls:
+  - hosts:
```
Please use a consistent list indentation style throughout the chart, either zero or two spaces.

Commit:
SWAGGER: remove from readme values that shouldn't be changed

keep them as templated values though, because if there is a change in the image used, the value only gets changed in one place instead of 8

Commit:
SWAGGER: fix indentation to be consistent

Commit:
SWAGGER: rename the key used for service port name to not be misleading

Commit:
SWAGGER: update swagger version to v3.14.1

Commit:
SWAGGER: remove option for internalPort because @unguiculus thinks this is unacceptable

unguiculus commented on stable/swagger/templates/ingress-oauth.yaml:
```diff
@@ -0,0 +1,27 @@
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+  annotations:
+    {{- range $key, $value := .Values.ingress.oauth.oauthAnnotations }}
```
Just make annotations generally configurable without being OAuth-specific.

unguiculus commented on stable/swagger/templates/ingress.yaml:
```diff
@@ -0,0 +1,37 @@
+    {{- range $key, $value := .Values.ingress.annotations }}
+    {{ $key }}: {{ $value | quote }}
+    {{- end }}
+    {{- if .Values.ingress.basicAuth.enabled }}
```
Remove this block. It's specific to nginx-ingress. Annotations are already configurable.

unguiculus commented on stable/swagger/templates/secrets.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if .Values.ingress.basicAuth.enabled -}}
```
This should not be part of the chart. It's not a feature of Swagger.

unguiculus commented on stable/swagger/values.yaml:
```diff
@@ -0,0 +1,63 @@
+  # - hosts:
+  #   - swagger.example.com
+  #   secretName: swagger-tls
+
```
Basic auth an OAuth stuff is specific to nginx-ingress. Make annotations generally configurable instead.
